### ISSUE TITLE
Bug : Minus number in the Not Affected Result

### ISSUE THREAD

**Issue by samnawi** [inasafe/inasafe#2969]
### Problem
I found some strange result in the current InaSAFE version.
There is a minus number in the _Not Affected_ 
I found it in some of IF that I've tested, like Volcano Point, Volcano Polygon, Tsunami Vector, Flood Vector. I guess it will occur in all of IF that produces _Not Affected_ Column 
![screenshot_34](https://cloud.githubusercontent.com/assets/8871958/16218379/5ec70d24-37a7-11e6-8843-da574e19b802.png)
![screenshot_35](https://cloud.githubusercontent.com/assets/8871958/16218380/5ecc296c-37a7-11e6-85a3-0d7c36464ea5.png)
![screenshot_36](https://cloud.githubusercontent.com/assets/8871958/16218381/5ecc6eae-37a7-11e6-8333-38c49d3fb100.png)

### Expected Behaviour
I've tried to run in 3.4.0, and I did not find a problem in here. So, I expect the result will show like this
![screenshot_33](https://cloud.githubusercontent.com/assets/8871958/16219618/cc5b5426-37b1-11e6-9895-063b7a1b141a.png)

### Environment
* InaSAFE : 3.4.1, 3.4.2, develop
* QGIS : 3.14.3
* OS : WIn7 64 bit
* Data used --> using some of sample data in http://data.inasafe.org/
  * Exposure : Ebolobo Volcano, Ebulobo Hazard Map, Jakarta FLood, Ambon Tsunami
  * Hazard : OSM Building
  * Aggregation : - 

### Proposed Solution
Make it happy

### CC
@Charlotte-Morgan @timlinux @ismailsunni @felix-yew @ivanbusthomi

**Comment by adissadis** [comment]
Found this problem also but in other computer (Dewi and Harry's laptop)  this problem not appear.

I'm using:
QGIS 2.14.3
InaSAFE 3.4.2
OS: Win 7 64bit



Data that i used:
Flood vector DKI Jakarta
Flood Raster DKI Jakarta
Building OSM polygon jakarta
District_DKI_Jakarta

**Comment by NyakudyaA** [comment]
Found the problem when running ash on buildings.

![not_affected](https://cloud.githubusercontent.com/assets/2510900/16297806/b0d85352-3932-11e6-9637-388d9be87c5d.png)

QGIS 2.14.3
Inasafe 3.5
OS ubuntu 16.04
Hazard: Ash raster
Exposure: Building from OSM

**Comment by ismailsunni** [comment]
hi @samnawi @adissadis and @NyakudyaA 
Thanks for the issue, I will work on it.

Perhaps, can you give the url of the data or more specific location where I can get the data? I often confuse which one the data that you used. I tried to find in data.inasafe.org, but I couldn't find them. The data are also zipped in the data.inasafe.org, make it more difficult to find the data that you use.

**Comment by Gustry** [comment]
Milestones 3.6 with a "current sprint" label ?
Check the message from Admire above, he is using the standard tests layers.

**Comment by Gustry** [comment]
BTW, I think we should definitely include tests about the impact table. For now, in each IF, we are testing only the impact layer.
I tried already here : https://github.com/inasafe/inasafe/commit/e855790da608f74312e0c7776f05160d61f84417

**Comment by ismailsunni** [comment]
@Gustry 
Sorry, mis-putting the milestones. Fixed now.

Yeah, actually, it was after I write my comment and talked to him after walking 2 meters.

re: test the impact table
agree.

**Comment by ismailsunni** [comment]
Hi all, in case you want to test it, this is my data set:
[2969.zip](https://github.com/inasafe/inasafe/files/330234/2969.zip)

**Comment by ismailsunni** [comment]
Hi @inasafe/developers @inasafe/issue-manager 
I have replicated this ticket with smaller area and more clear to see the problem.

Here my data test: 
Hazard :  [classified_hazard.zip](https://github.com/inasafe/inasafe/files/338664/classified_hazard.zip)
Exposure: 
[small_building.zip](https://github.com/inasafe/inasafe/files/338667/small_building.zip)

And if you run the analysis, you will get:
<img width="1385" alt="screen shot 2016-06-29 at 13 51 58" src="https://cloud.githubusercontent.com/assets/1421861/16443043/a692abc2-3e00-11e6-8346-431c585e5cd3.png">

Several problems we can see here:

1. See the number of building is 25 (index start from 0), but the total building is 3
2. If we excluded the building that outside the analysis extent, there should be : 25 - 5 (grey one) = 20 buildings.
3. The style is wrong for green. The hazard that has value = 4, 5, and 6 are classified as green or not affected. See the mapping below
<img width="315" alt="screen shot 2016-06-29 at 13 57 44" 
src="https://cloud.githubusercontent.com/assets/1421861/16443180/7889c200-3e01-11e6-8917-ffab54b872d7.png">

It seems the problem is assigning Not affected for hazard value = 4, 5, and 6.

**Comment by ismailsunni** [comment]
Problem found, the ClassifiedRasterHazardBuildingFunction doesn't use the hazard value map.

**Comment by timlinux** [comment]
Nice catch @ismailsunni !

**Comment by ismailsunni** [comment]
Btw, ClassifiedRasterHazardPopulationFunction also doesn't use hazard value map. I need to fix for both IFs.

**Comment by ismailsunni** [comment]
Update:
I have fixed it for ClassifiedRasterHazardBuildingFunction. It looks good now in QGIS. But, I got this itchy problem. See image below:
<img width="1378" alt="screen shot 2016-06-30 at 17 30 08" src="https://cloud.githubusercontent.com/assets/1421861/16485338/a299df48-3ee8-11e6-9a9e-1fd8a28d8d0d.png">

As you see, in the dock, the High hazard zone is 5 (red arrow), meanwhile, in the unit test (python console) it only gets 4 (green arrow). Obviously, it should be 5 buildings.

I have checked the building and I found out that the building that has label 4 (blue arrow) is the building that makes a difference. It is classified as high in InaSAFE QGIS, but Not affected in unit test.

You can try in your local using my inasafe branch : https://github.com/ismailsunni/inasafe/tree/fix_2969_minus_number

You only need to load classified_hazard.tif and small_building.shp, both are already included in safe/test/data/

cc @timlinux @Gustry @lucernae

**Comment by ismailsunni** [comment]
Ok, I fixed the bug that doesn't create the problem in the ticket.

I tried again my fix. Downloading from OSM building. The minus show up again. I try to use the data as a dataset. I saw some unclassified building TYPE. I classified it. Then the minus is gone when I ran the analysis.

Better experiment:
1. Download OSM building.
2. Make a copy of it.
3. Run keyword wizard, make a proper value mapping for the second layer.
4. Run analysis for both layers against any hazard.
5. The first one will give you minus number, the second one will show you the proper number.

In conclusion:
1. I fixed bug #2990, accidentally.
2. This bug hasn't been fixed since I modify the keyword, and make it works again.


For @samnawi @adissadis @NyakudyaA if you have the same problem, the short solution should be making sure the value mapping is correct. Or just simply, re-map it.

**Comment by ismailsunni** [comment]
Epilog:

Finally, I found the problem. There are two reason why this problem occurs:

1. *not proper* structure value mapping from OSM Downloader.
If you download building from OSM, you will get structure value mapping like this:
<img width="566" alt="screen shot 2016-07-01 at 08 23 54" src="https://cloud.githubusercontent.com/assets/1421861/16509276/3235f192-3f65-11e6-8cc6-bb6992c7c2cd.png">

See there is this `residence` and `economy`. Meanwhile, we have this list of known structure classes:
```python
[
'residential', 
'education', 
'health', 
'transport', 
'place of worship', 
'government', 
'commercial', 
'recreation', 
'public facility', 
'other'
]
```
In that list, there is no `residence` nor `economy`.
2. This line https://github.com/inasafe/inasafe/blob/version-3_4_2/safe/utilities/utilities.py#L154
In this line, it tries to re-order the dictionary based on `expected_key_order` (from the definitions or the structure classes). It assumes that the unordered has the same key as the expected_key_order.

The problem will occur if the layer contains some mapping to class that not exist in known classes, it will be _removed_ when InaSAFE try to reorder using the point number 2 above.

That's why, this ticket is solved by remapping the value map.

I have made a fix that make the unknown classes will still be included in the order. See https://github.com/inasafe/inasafe/pull/2999/commits/cb442cbf05b0f600868fc57af51ae12311ae4446

CC @samnawi @adissadis @timlinux @Gustry @NyakudyaA

**Comment by ismailsunni** [comment]
And for issue in my comment (https://github.com/inasafe/inasafe/issues/2969#issuecomment-229623471), I found out that impact_function.run() has different output as impact_function.run_analysis(). If I use the second one, it will give the same result as in the dock. I am still not sure why though. Need deeper checking.

**Comment by felix-yew** [comment]
Hi mas @ismailsunni 

I have tried also in InaSAFE develop 3.5, in a case of impact function (IF) earthquake (EQ) in West Sumatera 2009 with population data from AsiaPop. In this population IF, I use the 'ITB Fatality Function based on a Bayesian Approach', but I still find this bug in the 'Unaffected population' with the minus result. See the image below:
![image](https://cloud.githubusercontent.com/assets/13621886/17093608/79f7db40-5274-11e6-943d-f24eaa4a1568.png)

Hope, this bug can fix before the latest InaSAFE release.

Make it happy :smile:

**Comment by NyakudyaA** [comment]
@felix-yew  I have tested the same setup and I have no negative numbers in my results. 
OS: Ubuntu 16.04
Inasafe: 3.5
Data: Same as used by @felix-yew 

![padang](https://cloud.githubusercontent.com/assets/2510900/17326216/23a77966-58b0-11e6-8d24-0d1aeda11980.png)

**Comment by felix-yew** [comment]
Hi @NyakudyaA 

After I did a git pull in this morning and test again with IFCW, the result in 'Unaffected population' still negative. See the image below :
![image](https://cloud.githubusercontent.com/assets/13621886/17354501/3c16d014-5973-11e6-94e2-487585f69b31.png)

I have modified the analysis extent. 
![image](https://cloud.githubusercontent.com/assets/13621886/17354529/778faed6-5973-11e6-8aa4-1e3e62994711.png)

Maybe, if I modify the analysis extent the problem will appear. But, if I use full extent like this image :
![image](https://cloud.githubusercontent.com/assets/13621886/17354566/ba4d1402-5973-11e6-8514-f43dfd64673e.png)

The result is same like you've got.
![image](https://cloud.githubusercontent.com/assets/13621886/17354590/f74b1fac-5973-11e6-82c6-4fc7a89c95ce.png)

I think the problem in the analysis extent that we want.

OS : Win 7 64-bit/ 12 GB
InaSAFE 3.5 dev

Mas @ismailsunni this is the data :
[EQ Raster Hazard.zip](https://github.com/inasafe/inasafe/files/398608/EQ.Raster.Hazard.zip)

**Comment by ismailsunni** [comment]
Hi @felix-yew @NyakudyaA 
I checked this issue with the data from Felix. The issue occurs when you set custom extent and use ITB Bayesian model. It doesn't occur in ITB Fatality or Pager. I will check this.

**Comment by ismailsunni** [comment]
I will close this since the current minus number has different problem with the original one. See issue #3103 for tracking the minus number in ITB Bayesian IF.
